model.py

```diff
@@ -1,103 +1,1394 @@
+            vector = child.__set_parameter_vector__(vector)
+
+        # Now, add my own.
+        # Remember, backwards!
+        for name, val in zip(self.node_varpars[::-1], vector[::-1]):
+            par = getattr(self, name)
+            par.currVal = val
+
+        n_used = len(vector) - len(self.node_varpars)
+        return vector[:n_used]
+
+    def __check_par_assignments__(self):
+        '''Loop through my variables, and make sure that the Param.name is the
+        same as what I've got listed in self.node_par_names. This is probably
+        paranoid, but it makes me feel safer'''
+
+        param_dict = {key: getattr(self, key) for key in self.node_par_names}
+
+        for key, value in param_dict.items():
+            if key != value.name:
+                fail_msg = "Incorrect parameter name, {} assigned to {}. \nParameters are taken in the order {}".format(
+                    value.name, key, self.node_par_names
+                )
+                raise NameError(fail_msg)
+
+    # Properties to make everything cleaner
+    @property
+    def name(self):
+        '''The name of this object, of the form <class name>_<label>'''
+        return "{}_{}".format(self.__class__.__name__, self.label)
+
+    @property
+    def parent(self):
+        '''My parent <3'''
+        return self.__parent
+
+    @parent.setter
+    def parent(self, parent):
+        '''When setting the parent, I also need to add myself to their list of
+        children'''
+        self.__parent = parent
+        if self.__parent is None:
+            pass
+        else:
+            self.__parent.add_child(self)
+
+    @property
+    def children(self):
+        return self.__children
+
+    @children.setter
+    def children(self, children):
+        # I need to preserve the order of the children, so keep as a list.
+        if not isinstance(children, list):
+            children = list(children)
+
+        # Set the internal variable
+        self.__children = children
+
+        # Make sure my children know who's in charge
+        for child in self.__children:
+            child.__parent = self
+
+    @property
+    def dynasty_par_names(self):
+        '''A list of the keys to self.dynasty_par_vals'''
+        return self.__get_descendant_parameter_names__()
+
+    @property
+    def dynasty_par_vals(self):
+        '''A list of the variable parameter values below this node'''
+        return self.__get_descendant_parameter_vector__()
+
+    @dynasty_par_vals.setter
+    def dynasty_par_vals(self, dynasty_par_vals):
+        if not len(dynasty_par_vals) >= len(self.dynasty_par_vals):
+            raise ValueError('Wrong vector length on {} - Expected {}, got {}'.format(self.name, len(self.dynasty_par_vals), len(dynasty_par_vals)))
+        self.__set_parameter_vector__(dynasty_par_vals)
+
+    @property
+    def ancestor_param_dict(self):
+        '''A dict of the Param objects ABOVE! this node'''
+        return {key: val for key, val in
```

Comment:
Shouldn't this be an ```OrderedDict```?

Reply:
I've never come across OrderedDict before... It certainly can't hurt though.